File: src/lib/quoteContract.ts
import quoteContract from "../../shared/quote-contract.json";

type KisMarketDetail = {
  productType: string;
  label: string;
};

const US_MARKET_DETAILS: Record<string, KisMarketDetail> = {
  NAS: { productType: "512", label: "Nasdaq" },
  NYS: { productType: "513", label: "NYSE" },
  AMS: { productType: "529", label: "AMEX" },
};

export const KIS_DOMESTIC_MARKET_DIV_CODE = quoteContract.kis.domestic.market_div_code;
export const KIS_DOMESTIC_EXCHANGE_LABEL = quoteContract.kis.domestic.exchange_label;
export const QUOTE_CACHE_FRESH_SECONDS = quoteContract.quote_cache.fresh_seconds;
export const QUOTE_CACHE_STALE_SECONDS = quoteContract.quote_cache.stale_seconds;

export const KIS_US_MARKETS = quoteContract.kis.us_exchange_order.map((excd) => ({
  excd,
  ...(US_MARKET_DETAILS[excd] ?? { productType: "", label: excd }),
}));

File: services/market-data/src/quote_contract.rs
use std::{sync::OnceLock, time::Duration};

use serde::Deserialize;

const QUOTE_CONTRACT_JSON: &str = include_str!("../../../shared/quote-contract.json");

#[derive(Deserialize)]
struct QuoteContract {
    kis: KisContract,
    quote_cache: QuoteCacheContract,
}

#[derive(Deserialize)]
struct KisContract {
    domestic: DomesticContract,
    us_exchange_order: Vec<String>,
}

#[derive(Deserialize)]
struct DomesticContract {
    market_div_code: String,
    exchange_label: String,
}

#[derive(Deserialize)]
struct QuoteCacheContract {
    fresh_seconds: u64,
    stale_seconds: u64,
}

fn contract() -> &'static QuoteContract {
    static CONTRACT: OnceLock<QuoteContract> = OnceLock::new();
    CONTRACT.get_or_init(|| {
        serde_json::from_str(QUOTE_CONTRACT_JSON).expect("shared quote contract must be valid JSON")
    })
}

pub fn kis_domestic_market_div_code() -> &'static str {
    contract().kis.domestic.market_div_code.as_str()
}

pub fn kis_domestic_exchange_label() -> &'static str {
    contract().kis.domestic.exchange_label.as_str()
}

pub fn kis_us_exchange_order() -> &'static [String] {
    contract().kis.us_exchange_order.as_slice()
}

pub fn quote_fresh_ttl() -> Duration {
    Duration::from_secs(contract().quote_cache.fresh_seconds)
}

pub fn quote_stale_ttl() -> Duration {
    Duration::from_secs(contract().quote_cache.stale_seconds)
}
